pymc3/variational/svgd.py

```diff
@@ -0,0 +1,135 @@
+# Refactored by Thomas Wiecki
+# Original implementation by Qiang Liu and Dilin Wang: https://github.com/DartML/Stein-Variational-Gradient-Descent/blob/master/python/svgd.py
+# Many thanks to Dilin Wang for help with getting this going
+# (c) 2016 Qiang Liu and Dilin Wang
+
+import numpy as np
+import theano
+import theano.tensor as tt
+from tqdm import tqdm
+
+import pymc3 as pm
+from pymc3.model import modelcontext
+
+def vgd_kernel(X):
+    XY = tt.dot(X, X.transpose())
+    x2 = tt.reshape(tt.sum(tt.square(X), axis=1), (X.shape[0], 1))
+    X2e = tt.repeat(x2, X.shape[0], axis=1)
+    H = tt.sub(tt.add(X2e, X2e.transpose()), 2 * XY)
+
+    V = H.flatten()
+
+    # median distance
+    h = tt.switch(tt.eq((V.shape[0] % 2), 0),
+        # if even vector
+        tt.mean(tt.sort(V)[ ((V.shape[0] // 2) - 1) : ((V.shape[0] // 2) + 1) ]),
+        # if odd vector
+        tt.sort(V)[V.shape[0] // 2])
+
+    h = tt.sqrt(0.5 * h / tt.log(X.shape[0].astype('float32') + 1.0)) / 2.
+
+    Kxy = tt.exp(-H / h ** 2 / 2.0)
+    dxkxy = -tt.dot(Kxy, X)
+    sumkxy = tt.sum(Kxy, axis=1).dimshuffle(0, 'x')
+    dxkxy = tt.add(dxkxy, tt.mul(X, sumkxy)) / (h ** 2)
+
+    return (Kxy, dxkxy)
+
+
+def _svgd_run(x0, svgd_grad, n_iter=1000, stepsize=1e-3,
+              bandwidth=-1, alpha=0.9, progressbar=True):
+
+    theta = np.copy(x0)
+
+    # adagrad with momentum
+    fudge_factor = 1e-6
+    historical_grad = 0
+
+    if progressbar:
+        progress = tqdm(np.arange(n_iter))
+    else:
+        progress = np.arange(n_iter)
+
+    lnpgrad = np.empty_like(theta)
+    for i in progress:
+        grad_theta = svgd_grad(theta)[0]
+
+        # adagrad
+        if iter == 0:
+            historical_grad = historical_grad + grad_theta ** 2
+        else:
+            historical_grad = alpha * historical_grad + (1 - alpha) * (grad_theta ** 2)
+        adj_grad = np.divide(grad_theta, fudge_factor+np.sqrt(historical_grad))
+        theta = theta + stepsize * adj_grad
```

Comment:
Shouldn't the stepsize get smaller as the procedure iterates, or does it not matter?